Index: fstools-2018-01-04-37762ff4/mount_root.c
===================================================================
--- fstools-2018-01-04-37762ff4.orig/mount_root.c
+++ fstools-2018-01-04-37762ff4/mount_root.c
@@ -30,58 +30,22 @@
 static int
 start(int argc, char *argv[1])
 {
+	/*** SPACEX PATCH 1000. Always mount rootfs, ignore rootfs_data, mount /etc tmpfs's ***/
+
 	struct volume *root;
-	struct volume *data = volume_find("rootfs_data");
 	struct stat s;
 
 	if (!getenv("PREINIT") && stat("/tmp/.preinit", &s))
 		return -1;
 
-	if (!data) {
-		root = volume_find("rootfs");
-		volume_init(root);
-		ULOG_NOTE("mounting /dev/root\n");
-		mount("/dev/root", "/", NULL, MS_NOATIME | MS_REMOUNT, 0);
-	}
-
-	/*
-	 * Before trying to mount and use "rootfs_data" let's check if there is
-	 * extroot configured. Following call will handle reading config from
-	 * the "rootfs_data" on its own.
-	 */
-	extroot_prefix = "";
-	if (!mount_extroot()) {
-		ULOG_NOTE("switched to extroot\n");
-		return 0;
-	}
-
-	/* There isn't extroot, so just try to mount "rootfs_data" */
-	volume_init(data);
-	switch (volume_identify(data)) {
-	case FS_NONE:
-		ULOG_WARN("no usable overlay filesystem found, using tmpfs overlay\n");
-		return ramoverlay();
-
-	case FS_DEADCODE:
-		/*
-		 * Filesystem isn't ready yet and we are in the preinit, so we
-		 * can't afford waiting for it. Use tmpfs for now and handle it
-		 * properly in the "done" call.
-		 */
-		ULOG_NOTE("jffs2 not ready yet, using temporary tmpfs overlay\n");
-		return ramoverlay();
-
-	case FS_EXT4:
-	case FS_F2FS:
-	case FS_JFFS2:
-	case FS_UBIFS:
-		mount_overlay(data);
-		break;
-
-	case FS_SNAPSHOT:
-		mount_snapshot(data);
-		break;
-	}
+	root = volume_find("rootfs");
+	volume_init(root);
+	ULOG_NOTE("mounting /dev/root\n");
+	mount("/dev/root", "/", NULL, MS_NOATIME | MS_REMOUNT, 0);
+
+	mount("tmpfs", "/etc/config", "tmpfs", MS_NOATIME | MS_NOEXEC, "mode=0644");
+	mount("tmpfs", "/etc/wireless", "tmpfs", MS_NOATIME | MS_NOEXEC, "mode=0644");
+	mount("tmpfs", "/etc/map", "tmpfs", MS_NOATIME | MS_NOEXEC, "mode=0644");
 
 	return 0;
 }
@@ -101,24 +64,6 @@ stop(int argc, char *argv[1])
 static int
 done(int argc, char *argv[1])
 {
-	struct volume *v = volume_find("rootfs_data");
-
-	if (!v)
-		return -1;
-
-	switch (volume_identify(v)) {
-	case FS_NONE:
-	case FS_DEADCODE:
-		return jffs2_switch(v);
-
-	case FS_EXT4:
-	case FS_F2FS:
-	case FS_JFFS2:
-	case FS_UBIFS:
-		fs_state_set("/overlay", FS_STATE_READY);
-		break;
-	}
-
 	return 0;
 }
 