diff --git a/net/8021q/vlan_dev.c b/net/8021q/vlan_dev.c
index ac9791dd..d0c89d3e 100644
--- a/net/8021q/vlan_dev.c
+++ b/net/8021q/vlan_dev.c
@@ -36,6 +36,8 @@
 #include "vlanproc.h"
 #include <linux/if_vlan.h>
 #include <linux/netpoll.h>
+#include <linux/netfilter.h>
+#include <net/netfilter/nf_hnat.h>
 
 /*
  *	Create the VLAN header for an arbitrary protocol layer
@@ -758,6 +760,26 @@ static int vlan_dev_get_iflink(const struct net_device *dev)
 	return real_dev->ifindex;
 }
 
+static int vlan_dev_hnat_check(struct hnat_hw_path *path)
+{
+	struct net_device *dev = path->real_dev;
+	struct vlan_dev_priv *vlan = vlan_dev_priv(dev);
+
+	if (path->flags & HNAT_PATH_VLAN)
+		return -EEXIST;
+
+	path->flags |= HNAT_PATH_VLAN;
+	path->vlan_proto = vlan->vlan_proto;
+	path->vlan_id = vlan->vlan_id;
+	path->virt_dev = dev;
+	path->real_dev = vlan->real_dev;
+
+	if (vlan->real_dev->netdev_ops->ndo_hnat_check)
+		return vlan->real_dev->netdev_ops->ndo_hnat_check(path);
+
+	return 0;
+}
+
 static const struct ethtool_ops vlan_ethtool_ops = {
 	.get_settings	        = vlan_ethtool_get_settings,
 	.get_drvinfo	        = vlan_ethtool_get_drvinfo,
@@ -795,6 +817,7 @@ static const struct net_device_ops vlan_netdev_ops = {
 	.ndo_fix_features	= vlan_dev_fix_features,
 	.ndo_get_lock_subclass  = vlan_dev_get_lock_subclass,
 	.ndo_get_iflink		= vlan_dev_get_iflink,
+	.ndo_hnat_check		= vlan_dev_hnat_check,
 };
 
 static void vlan_dev_free(struct net_device *dev)
