diff -urN a/include/uapi/linux/if_ether.h b/include/uapi/linux/if_ether.h
--- a/include/uapi/linux/if_ether.h	2019-04-02 20:58:20.782345333 +0800
+++ b/include/uapi/linux/if_ether.h	2019-04-02 20:58:28.570165386 +0800
@@ -22,6 +22,7 @@
 #define _UAPI_LINUX_IF_ETHER_H
 
 #include <linux/types.h>
+#include <linux/libc-compat.h>
 
 /*
  *	IEEE 802.3 Ethernet magic constants.  The frame sizes omit the preamble
diff -urN a/libc-compat.h b/libc-compat.h
--- a/include/uapi/linux/libc-compat.h	2019-04-02 20:58:45.321776594 +0800
+++ b/include/uapi/linux/libc-compat.h	2019-04-02 20:58:53.281591046 +0800
@@ -85,6 +85,14 @@
 
 #endif /* _NET_IF_H */
 
+/* musl defines the ethhdr struct itself in its netinet/if_ether.h.
+ * Glibc just includes the kernel header and uses a different guard. */
+#if defined(_NETINET_IF_ETHER_H)
+#define __UAPI_DEF_ETHHDR              0
+#else
+#define __UAPI_DEF_ETHHDR              1
+#endif
+
 /* Coordinate with libc netinet/in.h header. */
 #if defined(_NETINET_IN_H)
 
@@ -161,6 +169,9 @@
 /* For the future if glibc adds IFF_LOWER_UP, IFF_DORMANT and IFF_ECHO */
 #define __UAPI_DEF_IF_NET_DEVICE_FLAGS_LOWER_UP_DORMANT_ECHO 1
 
+/* Definitions for if_ether.h */
+#define __UAPI_DEF_ETHHDR              1
+
 /* Definitions for in.h */
 #define __UAPI_DEF_IN_ADDR		1
 #define __UAPI_DEF_IN_IPPROTO		1
