diff --git a/arch/arm/boot/dts/mt7623.dtsi b/arch/arm/boot/dts/mt7623.dtsi
index 21a6ea6f..4534da80 100644
--- a/arch/arm/boot/dts/mt7623.dtsi
+++ b/arch/arm/boot/dts/mt7623.dtsi
@@ -1227,7 +1227,7 @@
 	};
 
         hnat: hnat@1b000000 {
-                compatible = "mediatek,mtk-hnat";
+                compatible = "mediatek,mtk-hnat_v1";
                 reg = <0 0x1b100000 0 0x3000>;
                 resets = <&ethsys 0>;
                 reset-names = "mtketh";
